### PR TITLE
Fix mirror updates when adding new reqs and dev-reqs are present

### LINKED ISSUE #9514: composer2 can't find package versions but composer1 finds it fine
My `composer.json`:

```json
{
    "name": "jacques/php-payfast-parsers",
    "description": "Parsers for PayFast Reconciliation Files",
    "repositories": [
    ],
    "config": {
        "optimize-autoloader": true,
        "sort-packages": true
    },
    "require": {
        "league/csv": "^9.6",
        "moneyphp/money": "^3.3"
    },
    "require-dev": {
        "friendsofphp/php-cs-fixer": "^2.16",
        "phpmd/phpmd": "^2.8",
        "phpro/grumphp": "^0.18",
        "phpunit/php-invoker": "^3.1",
        "phpunit/phpunit": "^9.4",
        "rector/rector": "^0.8.6",
        "squizlabs/php_codesniffer": "^3.5",
        "vimeo/psalm": "^4.1"
    },
    "autoload": {
        "psr-4": {
            "Jacques\\PayFast\\Parsers\\": "src/"
        }
    },
    "autoload-dev": {
        "psr-4": {
            "Jacques\\PayFast\\Parsers\\Tests\\": "tests/"
        }
    },
    "license": "mpl",
    "authors": [
        {
            "name": "Jacques Marneweck",
            "email": "jacques@powertrip.co.za"
        }
    ],
    "scripts": {
        "test": "vendor/bin/phpunit --colors --debug --configuration=tests/phpunit.xml --coverage-clover build/logs/clover.xml --coverage-text"
    }
}
```

[Commit after using composer on an older v1.x build](https://github.com/jacques/php-payfast-parsers/commit/d22019e3454a64f9f7cf0d270096f59e045dc181).

Output of `composer diagnose`:

```
23:10 $ composer diagnose
Checking composer.json: WARNING
License "mpl" is not a valid SPDX license identifier, see https://spdx.org/licenses/ if you use an open license.
If the software is closed-source, you may use "proprietary" as license.
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com oauth access: OK
Checking disk free space: OK
Checking pubkeys: 
Tags Public Key Fingerprint: 57815BA2 7E54DC31 7ECC7CC5 573090D0  87719BA6 8F3BB723 4E5D42D0 84A14642
Dev Public Key Fingerprint: 4AC45767 E5EC2265 2F0C1167 CBBB8A2B  0C708369 153E328C AD90147D AFE50952
OK
Checking composer version: OK
Composer version: 2.0.7
PHP version: 7.4.12
PHP binary path: /usr/local/Cellar/php/7.4.12/bin/php
OpenSSL version: OpenSSL 1.1.1h  22 Sep 2020
cURL version: 7.73.0 libz 1.2.11 ssl OpenSSL/1.1.1h
zip extension: OK
```

When I run this command:

```
composer update --lock
```

I get the following output:

```
Loading composer repositories with package information
Updating dependencies
Unable to find a compatible set of packages based on your non-dev requirements alone.
Your requirements can be resolved successfully when require-dev packages are present.
You may need to move packages from require-dev or some of their dependencies to require.

  Problem 1
    - Root composer.json requires league/csv ^9.6, found league/csv[9.5.0] but it does not match the constraint.
```

And I expected this to happen:

I would expect league/csv to update from 9.5 to 9.6 as the package exists.  On composer v1 on a SmartOS box this updated fine.

### PR DESCRIPTION
**Pull request by Seldaek**
Fixes #9514

### CODE REVIEW

File: src/Composer/Installer.php [hunk 1]
```diff
@@ -851,6 +832,33 @@ private function createRequest(RootPackageInterface $rootPackage, PlatformReposi
         return $request;
     }
 
+    private function requirePackagesForUpdate(Request $request, LockArrayRepository $lockedRepository = null, $includeDevRequires = true)
+    {
+        // if we're updating mirrors we want to keep exactly the same versions installed which are in the lock file, but we want current remote metadata
+        if ($this->updateMirrors) {
+            $excludedPackages = array();
+            if (!$includeDevRequires) {
+                $excludedPackages = $this->locker->getDevPackageNames();
```
Comment (naderman) [inline src/Composer/Installer.php:841]: Could use array_flip and isset below to avoid searching through the list for each package?

**Review by naderman — APPROVED** [review]

File: src/Composer/Installer.php [hunk 2]
```diff
@@ -820,7 +801,7 @@ private function createPolicy($forUpdate)
      * @param  RepositoryInterface|null $lockedRepository
```
Comment (stof) [inline src/Composer/Installer.php:801]: this should be updated to match the typehint

Comment (Seldaek) [reply src/Composer/Installer.php:801]: We have phpstan for this (see below) so you don't need to type check ;)

File: src/Composer/Installer.php (hunk 1 above)
Comment (Seldaek) [reply src/Composer/Installer.php:841]: Kinda micro-optimization level here given the amount of packages involved, but sure will do

Comment (naderman) [reply src/Composer/Installer.php]: Sure, but it's not really any more code ;-) Thanks

**Pull request merged**